src/main/java/com/powsybl/openloadflow/equations/EquationSystem.java

```diff
@@ -186,6 +188,14 @@ private Equation<V, E> addEquation(Pair<Integer, E> p) {
         return equation;
     }
 
+    private Equation<V, E> newEquation(int num, E type) {
```

Comment:
This is really a terrible hack. It breaks the genericity if EquationSystem that is not supposed to know equations type (this is the purpose of the generic type). It creates bad cyclic dependencies between modules.

Not clear to me even at looking the issue to figure what really is the bug. Could you try to create a clear test case that show the issue (and not changing an existing one)? 

Anyway even if we really have an issue I'm sure we can do better to fix it.

Reply:
Correct me @vmouradian if I am wrong or if I forget something.

The issue is that, when you have a non impedant branch connected to a bus, you should recompute the flow on this branch. Surprisingly, we do not use the dummy_p variable value, instead we use the balance of the bus (calling bus.getP())
Flow on the branch = bus (im)balance = - sum (flows on other bus terminals)

When you have a PST on one branch connected to this bus, the flow generated by the shifting angle is not taken into account in DC when you call bus.getP(). -> this is the bug.

Reply:
> Surprisingly, we do not use the dummy_p variable value, instead we use the balance of the bus (calling bus.getP())
> Flow on the branch = bus (im)balance = - sum (flows on other bus terminals)

So this is a hack on a hack ? :-) 
I think using dummy_p for getting flow on a branch only work for an isolated non impedant branch. When there is mutliple non impedant branches connected we have to rely on `ZeroImpedanceFlows`

Reply:
> When you have a PST on one branch connected to this bus, the flow generated by the shifting angle is not taken into account in DC when you call bus.getP(). -> this is the bug.

I don't understand where is the bug in bus.getP(), we already take into account the RHS containing the phase shifting part.

Reply:
@vmouradian please:
- revert the changes on the existing tests
- add new tests showing in which case the issue happens

and then it will be much easier to discuss on the solution

Reply:
I think that a better and easier method is to code an eval method in AbstractClosedBranchDcFlowEquationTerm without calling the "calculateSensi" one. The method should only compute ph2 - ph1 (or the opposite) times the "power" attribute.

Reply:
Indeed the eval function was not supposed to take into account the RHS (shift) part. This is probably a regression introduced by sensi calculation. So let's fix the real problem.

Reply:
> I think that a better and easier method is to code an eval method in AbstractClosedBranchDcFlowEquationTerm without calling the "calculateSensi" one. The method should only compute ph2 - ph1 (or the opposite) times the "power" attribute.

We can just add an option to calculateSensi to take into account the a1 variable or not depending on the context and call it without in the eval method.

Reply:
If we just change the value given by AbstractClosedBranchDcFlowEquationTerm.eval() to (phi2 - ph1) / X,  this will fix the issue for the lfBus.getP(), but then the value for active power on pst branches will be wrong. (lfbranch.getP().eval() will be wrong -> not taking A2 into account)

Reply:
What if we just consider that the eval() method of an equation term should contain the rhs part ? 
If we do so we simply need to remove [these lines](https://github.com/powsybl/powsybl-open-loadflow/blob/8822d5f16d09de29e7f3034650a1ec625894bd95/src/main/java/com/powsybl/openloadflow/equations/Equation.java#L159-L161) to fix the bug

Reply:
> What if we just consider that the result of eval() method for an equation term should contain the rhs part ? If we do so, we simply need to remove [these lines](https://github.com/powsybl/powsybl-open-loadflow/blob/8822d5f16d09de29e7f3034650a1ec625894bd95/src/main/java/com/powsybl/openloadflow/equations/Equation.java#L159-L161) to fix the bug +using the eval() like that is what is currently done in practice

I am not qualified enough to tell, but I can tell you that while OLF unit tests would be happy, on real test cases I get less imbalance KO but still many KO-s.

Reply:
> What if we just consider that the result of eval() method for an equation term should contain the rhs part ? If we do so, we simply need to remove [these lines](https://github.com/powsybl/powsybl-open-loadflow/blob/8822d5f16d09de29e7f3034650a1ec625894bd95/src/main/java/com/powsybl/openloadflow/equations/Equation.java#L159-L161) to fix the bug +using the eval() like that is what is currently done in practice

You can't do that in DC because there is an assumption by the DC engine that eval is linear, not affine.

Reply:
Maybe the solution is in DCEquationSystemCreator. For an non impedant branch, you may need to transfer the BUS_TARGET_P of one side to the other side. (maybe once all equations are created to be sure the get all RHS defined).

Reply:
> You can't do that in DC because there is an assumption by the DC engine that eval is linear, not affine.

Ok thanks, I was not aware of that. Do you know where we can see this ?

Reply:
> Maybe the solution is in DCEquationSystemCreator. For an non impedant branch, you may need to transfer the BUS_TARGET_P of one side to the other side. (**edit** - should be done in a way that works with securty analysis - disconnection of the element bringing the RHS - disconnection of the zero impedance line or one of its bus etc...)

I do not realy understand. The issue is that the eval() of the BUS_TARGET_P (or any other)  equations is wrong if one of the terms has rhs in his eval() (like pst or hvdc in DC). 
In my comprehension, your suggestion is transferring the equation from one side of the NI branch to the other, and still in my comprehension this should not solve the issue. Does it ?

Reply:
If I remember well, the DC engine doesn't call eval. SO any constant term has t be in the RHS.
Anyway, since this goes deep into the equation system we need a better coverage for this change thatn includes AC tests, sensi tests etc...
I wil try to spend time on this newt week.
For me the ideal solution would only touch the zero impedance equation logic.

Reply:
There is a contract in the equation system that says eval = rhs + target vector
If we break this contract we do not know the effect.
This is why I would like to be sure we fully understand what's going on.

Reply:
For the moment i can suggest 3 solutions : 
- Remove the rhs lines in the Equation::eval()  (reset to [9d5a6ec](https://github.com/powsybl/powsybl-open-loadflow/pull/1138/commits/9d5a6ec04ea4cb85ca75b38589ac22d3af6941dc))
- Not count the rhs for Impedant branches in ZeroImpedanceFlows (see [e70857d](https://github.com/powsybl/powsybl-open-loadflow/pull/1138/commits/e70857d8154f2c1767a966edec5494a5e7ecd5ab)
- Make equation terms' eval() affine and create intermediary class that can give the equation and compute the flow with rhs() (will dev it soon)

Reply:
I start to think that there are some confusions in the current equation system object model.

In the current model, an equation term represent a physical quantity.
The physical quantity is (in the current confusing model):   eval() - rhs()
   This is what is used in the eval implementation of an equation that returns the pysical quantity 

But for the DC engine, we must explicitely separate the linear part from the constant part so taht we can rite

TARGET vector - RHS vector = MATRIX*(vector of V and phi)
And this separation is done by using for a term:
   eval as the linear part
   rhs as the constant part

On the other hand, most of the code uses eval of a term to get a physical quantity. But the pysical quantity is, with this model  term.eval - term.rhs

I'm still not sure how to fix that but I would tend to something like;
    eval returns the physical quantity for both equations (as it does today) and the terms
    create an interface for DC terms that force to define the linear part and the constant part
           use this to create the DC target vector and the DC matrix
           and use this for eval that should return the physical quantity

Reply:
> For the moment i can suggest 3 solutions :
> 
>     * Remove the rhs lines in the Equation::eval()  (reset to [9d5a6ec](https://github.com/powsybl/powsybl-open-loadflow/pull/1138/commits/9d5a6ec04ea4cb85ca75b38589ac22d3af6941dc))
> 
>     * Not count the rhs for Impedant branches in ZeroImpedanceFlows (see [e70857d](https://github.com/powsybl/powsybl-open-loadflow/pull/1138/commits/e70857d8154f2c1767a966edec5494a5e7ecd5ab))
> 
>     * Make equation terms' eval() affine and create intermediary class that can give the equation and compute the flow with rhs() (will dev it soon)

I will try another one:
    implement evallhs for use in the equation solver that wants to have   lhs = rhs + target
    eval that returns the full quantity
I'll let you know if it leads me somewhere. I think that currently the mangement of rhs is inconsistent in many places.

Reply:
@vmouradian @geofjamg  I created a proposal here https://github.com/powsybl/powsybl-open-loadflow/pull/1147
I think it clairifes the issue with RHS.

Reply:
> @vmouradian @geofjamg I created a proposal here https://github.com/powsybl/powsybl-open-loadflow/pull/1147
Thank you ! I is indeed a clearer and cleaner way to fix the problem

Reply:
Thanks. I merged @geofjamg  the code is now here and can be compared with main.